### ISSUE TITLE
Support @immutable

### ISSUE THREAD

**Issue by alexeieleusis** [dart-lang/linter#636]
I would like to have an annotation with static analysis that indicates that a class is immutable or a function is side effect free. This should be accompanied by a few lints:

- If a class has the annotation, warn if a field is not final.
- If a class does not have the annotation but all fields are final, suggest adding the annotation.
- If a function/method has the annotation, warn if it invokes something with `void` return type.
- If a function/method does not have the annotation and does not invoke anything returning `void` suggest to add the annotation, we need some other criteria to avoid suggesting annotating the whole world.

I understand it is very difficult to get this analysis right, but I think there are good use cases for this. More than a feature request this is to start a discussion to understand how viable it is.

**Comment by bwilkerson** [comment]
> If a class has the annotation, warn if a field is not final.

There is an annotation defined in the `meta` package that does the first of those four asks:
```
/// Used to annotate a class `C`. Indicates that `C` and all subtypes of `C`
/// must be immutable.
///
/// A class is immutable if all of the instance fields of the class, whether
/// defined directly or inherited, are `final`.
///
/// Tools, such as the analyzer, can provide feedback if
/// * the annotation is associated with anything other than a class, or
/// * a class that has this annotation or extends, implements or mixes in a
///   class that has this annotation is not immutable.
const Immutable immutable = const Immutable();
```

> If a class does not have the annotation but all fields are final, suggest adding the annotation.

We don't currently suggest adding the annotation because there isn't any way for us to know that the author of the class wants to prevent subclasses from adding mutable state.

> If a function/method has the annotation, warn if it invokes something with `void `return type.

I don't understand what the invocation of a `void` method has to do with immutability. Can you explain?

**Comment by alexeieleusis** [comment]
> I don't understand what the invocation of a void method has to do with immutability. Can you explain?

I am not sure myself, but I am thinking about making side effects explicit and invoking something with `void` return type is all about it's  side effects and the consumer might not have visibility into it.

It is very nice that a hint triggers if a class without the annotation extends another that has it.

**Comment by bwilkerson** [comment]
While it's true that `void` methods generally have side-effects, the converse isn't true. Just because a method returns a value doesn't mean that it doesn't have a side-effect. Which means that the `voidness` of methods isn't a reliable measure of whether they have side-effects.

The method issue sounds orthogonal to the immutability question, which means I'd suggest a second annotation, perhaps something like `@sideEffectFree` with the semantics that side-effect free functions can only invoke other side-effect free functions. Unfortunately, that would require developers to do a lot of work to annotate their functions, and you run into a problem when you get to the SDK. Not sure whether it provides enough value to developers to justify the cost.

**Comment by alexeieleusis** [comment]
There is one use case I would like to support, in this class I would like to lint if `S` is not immutable. 

```dart
/// Lint if S does not have the @immutable
class Something<S> {
  S get myS => ...;
}
```

Definitely beyond the scope of the annotation and the related rules, but it would be useful to communicate consumers it is preferred. Ideas?

Concrete example here https://github.com/alexeieleusis/greencat/blob/master/lib/src/store.dart#L43

**Comment by bwilkerson** [comment]
So you want to annotate the type parameter? What would the semantics of doing that be? (What hints, if any, would the analyzer produce as a result of seeing that annotation on a type parameter?)

**Comment by alexeieleusis** [comment]
I am not sure about semantics, but at instance creation, if the specified parameter is not annotated, lint:

```dart
class NotAnnotated {
...
}

@immutable
class Annotated {
...
}

void foo() {
  ...
  var something = new Something<NotAnnotated>(); // LINT
  var ok = new Something<Annotated>(); // OK
  ...
}
```

**Comment by bwilkerson** [comment]
Actually, I think you do know the semantics :-), or at least part of them. You want an annotation that can be applied to a type parameter that will cause a hint to be generated if a type argument is specified that is not annotated as being `immutable`. It would presumably also want to produce a hint if the type parameter's bound is not immutable.

The next question is how often such an annotation would be useful (and hence whether it's worth the complexity of adding it).

**Comment by alexeieleusis** [comment]
tl;dr Developers coming to Flutter from React/Redux will be used to this pattern, some of them will find the lint useful.

I wrote greencat (the library in the example) after learning that Flutter has some similarities with React in how business logic should be organized in an app and also learning that Redux is (back when I wrote greencat) the most popular library for that end, greencat is a port of Redux to dart. All of this to say that if Flutter attracts React/Redux developers, regardless of which library they use, (a good share) will want to have immutable objects for their stores.

**Comment by matanlurey** [comment]
Isn't this basically impossible? Consider:

```dart
@immutable
class Foo {
  // This is legal, but there is no way to know that statically with out lots of false positives.
  List<String> get foo => new ImmutableList([...]);
}

```dart
abstract class ImmutableList<T> implements List<T> {
  factory ImmutableList(values) = _ImmutableListImpl;
}
```

Dart's interfaces and abstract classes, which are an extremely popular convention, wouldn't be annotated with `@immutable`, so how would you know if the concrete implementation is indeed immutable or not?

**Comment by alexeieleusis** [comment]
Good point, we can have a blacklist of mutable classes in which `List` might be the first and honorary member.

But a very good point indeed.

**Comment by alexeieleusis** [comment]
Though the example you provide happens with what is already available, right? Asides the nit that the annotation works with fields and not getters, your example is not mutating the class itself.
